### PR TITLE
Fixed towncrier template rendering

### LINKED ISSUE #9102: Towncrier rendering
I checked the rendering and it needs a custom jinja template to render correctly.
The current template https://github.com/numba/numba/blob/main/docs/source/release/template.rst is not valid.
It should be something like https://github.com/numpy/numpy/blob/main/doc/release/upcoming_changes/template.rst?plain=1

### PR DESCRIPTION
**Pull request by kc611**
Resolves #9102,

As per the issue, 
Adds a template for towncrier's auto-release notes generation and fixes towncrier build rendering issues.

### CODE REVIEW

**Comment by gmarkall** [comment]
Moved into the 0.58RC milestone as it fixes #9102 which is in it.

File: docs/upcoming_changes/template.rst [hunk 1]
```diff
@@ -0,0 +1,39 @@
...
+{{ title }}
+{{ "=" * title|length }}
+
+{% for section, _ in sections.items() %}
+{% set underline = underlines[0] %}{% if section %}{{ section }}
+{{ underline * section|length }}{% set underline = underlines[1] %}
+
+{% endif %}
+{% if sections[section] %}
+{% for category, val in definitions.items() if category in sections[section] %}
```
Comment (sklam) [inline docs/upcoming_changes/template.rst:12]: We should change the ordering of the sections. Currently, the CUDA session is going before the new_features. I found this the ordering is driven by the sections ordering in towncrier.toml. I suggest the following changes to it:

```

[[tool.towncrier.type]]
    directory = "highlight"
    name = "Highlights"
    showcontent = true

[[tool.towncrier.type]]
    directory = "new_feature"
    name = "New Features"
    showcontent = true

[[tool.towncrier.type]]
    directory = "improvement"
    name = "Improvements"
    showcontent = true

[[tool.towncrier.type]]
    directory = "np_support"
    name = "NumPy Support"
    showcontent = true

[[tool.towncrier.type]]
    directory = "cuda"
    name = "CUDA API Changes"
    showcontent = true

[[tool.towncrier.type]]
    directory = "performance"
    name = "Performance Improvements and Changes"
    showcontent = true

[[tool.towncrier.type]]
    directory = "bug_fix"
    name = "Bug Fixes"
    showcontent = true

[[tool.towncrier.type]]
    directory = "change"
    name = "Changes"
    showcontent = true

[[tool.towncrier.type]]
    directory = "deprecation"
    name = "Deprecations"
    showcontent = true

[[tool.towncrier.type]]
    directory = "expired"
    name = "Expired Deprecations"
    showcontent = true

[[tool.towncrier.type]]
    directory = "compatibility"
    name = "Compatibility Notes"
    showcontent = true

[[tool.towncrier.type]]
    directory = "doc"
    name = "Documentation Changes"
    showcontent = true

[[tool.towncrier.type]]
    directory = "infrastructure"
    name = "Infrastructure Related Changes"
    showcontent = true
```

**Review by sklam — CHANGES_REQUESTED** [review]
Thanks for the patch. I noticed a few things that need changing:

- heading level is not aligning with previous CHANGE_LOG entries
- the issue link not going to numba repo
- the section ordering can be improved

**Review by esc — CHANGES_REQUESTED** [review]
for a search and replace mistake

File: docs/upcoming_changes/9095.bug_fix.rst [hunk 2]
```diff
@@ -1,5 +1,5 @@
-Support `dtype` keyword argument in `numpy.arange` with `parallel=True`
-=======================================================================
+Support `dtype` keyword argument in `numpy.arange` with `parallel"True`
```
Comment (esc) [inline docs/upcoming_changes/9095.bug_fix.rst:1]:
```suggestion
Support `dtype` keyword argument in `numpy.arange` with `parallel=True`
```

**Review by esc — CHANGES_REQUESTED** [review]
fixup missing `=`

**Comment by esc** [comment]
Testing via #9135 yields:

Screen Shot 2023-08-16 at 11 17 57

**Review by esc — APPROVED** [review]
I have verified via #9135 that the new headings look good and the `towncrier build --draft` output can be used to partially auto-generate the new Numba change log.

**Pull request merged**